launcher/tests/test_notify.py

```diff
@@ -0,0 +1,247 @@
+UPDATER_WARNING_REGEX = r"^Last successful update \(.* hours ago\) is above warning threshold "
+r"\(.* hours\). Uptime grace period of .* hours has elapsed (uptime: .* hours)."
+
+# Regex for info log if we've updated too long ago, but grace period still ticking
+GRACE_PERIOD_REGEX = r"Last successful update \(.* hours ago\) is above "
+r"warning threshold \(.* hours\). Uptime grace period of .* hours has not elapsed "
+r"yet \(uptime: .* hours\)."
+
+# Regex for info log if we've updated recently enough
+NO_WARNING_REGEX = r"Last successful update \(.* hours ago\) is below the warning threshold "
+r"\(.* hours\)."
+
+BAD_TIMESTAMP_REGEX = r"Data in .* not in the expected format."
+
+
+@mock.patch("Notify.sdlog.error")
+@mock.patch("Notify.sdlog.warning")
+@mock.patch("Notify.sdlog.info")
+def test_notify_lock(
+        mocked_info, mocked_warning, mocked_error
+):
+    """
+    Test whether we can successfully obtain an exclusive lock for the notifier
+    script
+    """
+    with TemporaryDirectory() as tmpdir:
+        notify.LOCK_FILE_NOTIFIER = os.path.join(tmpdir, "sdw-launcher.lock")
+        pid = os.getpid()
+        lh = notify.obtain_notify_lock()  # noqa: F841
+        # No handled exception should occur
+        assert not mocked_error.called
+        # We should be getting a lock handle back
+        assert lh is not None
+
+        cmd = ['lsof', '-w', notify.LOCK_FILE_NOTIFIER]
+        output_lines = subprocess.check_output(cmd).decode("utf-8").strip().split('\n')
+        # We expect exactly one process to be accessing this file, plus output header
+        assert len(output_lines) == 2
+        lsof_data = output_lines[1].split()
+        # We expect the output to have the standard number of fields
+        assert len(lsof_data) == 9
+        # We expect the PID column to contain the ID of this process
+        assert lsof_data[1] == str(pid)
+        # We expect an exclusive write lock to be set for this process
+        assert lsof_data[3].find('W') != -1
+
+
+@mock.patch("Notify.sdlog.error")
+@mock.patch("Notify.sdlog.warning")
+@mock.patch("Notify.sdlog.info")
+def test_updater_lock_prevents_notifier(
+    mocked_info, mocked_warning, mocked_error
+):
+    """
+    Test whether an exlusive lock on the updater lock file prevents the notifier
+    from launching (so it does not come up when the user is in the process of
+    updating).
+    """
+    with TemporaryDirectory() as tmpdir:
+        notify.LOCK_FILE_LAUNCHER = os.path.join(tmpdir, "sdw-launcher.lock")
+        updater.LOCK_FILE = os.path.join(tmpdir, "sdw-launcher.lock")
+        lh = updater.obtain_lock()  # noqa: F841
+
+        # We're in the same process, so obtaining an additional lock would
+        # always succeed. We use the multiprocessing module to run the function
+        # as a separate process.
+        p = Pool(processes=1)
+        lock_result = p.apply(notify.can_obtain_updater_lock)
+        p.close()
+        assert lock_result is False
```

Comment:
I've struggled to find a clean way to provoke a lock conflict in a test that works well across platforms. Pointers appreciated here as well. This method does appear to work, but as far as I can tell, it causes the code coverage for those lines to be misreported.

Reply:
I think here it would be fine to mock `open` or `lockf` and have `IOException` (not tested, but as example, using the following decorators :

```@mock.patch("builtins.open", side_effect=IOError())```
and 
```@mock.patch("fcntl.lockf", side_effect=IOError())``` 

depending on which part of the lock logic you want to test